tools/smascii.1
.TH SMASCII "1" "March 2010" "help2man and hand edition" "User Commands"
.SH NAME
smascii \- smascii
.SH SYNOPSIS
.B smascii
\fIsmaky_file converted_file \fR[\fI--unix\fR]
.SH DESCRIPTION
Tool to convert Smaky text file to Extended ASCII (ISO\-8859\-1).
.TP
smaky_file
the smaky text file
.TP
converted_file
the file converted
.TP
\fB\-\-unix\fR
the Carriage Return (Old Mac) will be
converted to Line Feed (unix)
.SH BUGS
.PP
Please, report bugs to <mathieu@schroetersa.ch>.
.SH AUTHOR
.PP
This manpage was written for Debian, starting with a help2man output, by
Didier Raboud <didier@raboud.com>, on the 4. March 2010.


ChangeLog
2023-03-26  Mathieu Schroeter <mathieu@schroetersa.ch>

	* Release 1.0.0

	* fosdd: new tool that can be used in order to convert floppy device
	  or image into hard disk image and vice-versa.

	* fosmount: the PBM and XPM images use a shorter filename.

2010-03-14  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Release 0.4.0

	* Many improvements in the build scripts and Makefiles. The build for
	  Windows uses the --cross-compile option with the configure script.
	  The external scripts to build for Windows, provided with the older
	  versions are no longer necessary. Now the DLLs are created in the
	  right way.
	  The pkg-config files are now available for libfosfat and libfosgra.

	* libfosgra: add support for the decoding of .COLOR files.

	* fosmount: add the conversion on the fly of .COLOR files to .XPM files.
	  Note that the XPM files use the version 2 of X PixMap. It needs a real
	  application like The Gimp in order to read the files. Or the tool SXPM
	  from the X11 (X.Org) project.

	* smascii: add a manpage. Thanks to Didier Raboud from Debian.

	* libw32disk: this library doesn't rely anymore to C++. It is now fully
	  rewritten in C with a proper license in order to be redistribuable by
	  Debian.

2009-07-07  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* New beta released 0.3.2

	* libfosgra: new library to support .IMAGE decoding.

	* fosmount: some improvements to convert on the fly the images [.IMAGE]
	  to PBM (portable bitmap format) with libfosgra.

	* Fix doxygen, packaging and Makefiles.

2009-07-04  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* New beta released 0.3.1

2009-05-06  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* fosrec: Add a new tool to easily recovery files.

	* Fix Debian/Ubuntu packaging.

2009-01-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* A lot of cleaning and all doxygen documentations are only in the
	  public header file.

	* Add option in configure to generate doxygen HTML and PDF outputs.

2008-02-04  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Improve a lot the code lisibility.

	* Fix some memleaks when errors appear.

2008-01-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Add manpages for fosmount and fosread.

2007-12-16  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* New major beta released 0.3.0

	* Add Window$ support, and use MinGW32 to compile DLL libraries,
	  fosread and smascii. A script package-win32.sh can be used to create
	  a zip package.

	* libfosfat: add support for Window$ through libw32disk.

	* libw32disk: add C wrapper to use original Disk.cxx class with
	  libfosfat.

	* fosread: fix symlink support, now that is possible to read like a
	  directory. Add support to list and restore deleted files.

2007-12-13  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: first implementation of an undelete tool just disabled by
	  default. The next step is added some options to list and restore
	  files.

2007-12-12  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: add support for multiple devices. A new main structure is
	  used instead of global variables. Only the logger is always global.

2007-12-09  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Fourth beta released 0.2.3

	* Some fixes, cleaned code, ...

2007-07-13  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Change license GPLv2 to GPLv3.

	* Fix a lot of memory leaks.

2007-07-07  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Third beta released 0.2.2

2007-06-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: change the auto detection function for a better way. Use
	  the name was very bad because a lot of disks have no name. Then the
	  test is does with the CHK (look the code for more informations).

2007-06-24  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Add auto detection for the disk type. The type can be forced with
	  fosmount and fosread if necessary.

	* libfosfat: an internal logger can be enabled to know what the FOS is
	  doing. A new option is available in fosread and fosmount.

2007-06-22  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Second beta released 0.2.1

	* Add support for symbolic links. The Smaky's soft links are now
	  converted to symlink.

	* libfosfat: the cache is now released when the device is closed.

2007-06-20  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: a new system of caching is implemented to improve the
	  speed. Now that will be very fast for looking in a directory. The
	  system must be fixed for that the memory is released.

	* fosmount: a bug is fixed for that FUSE is always loaded as
	  single-thread. If not there is a really a big problem when data are
	  loaded with libfosfat.

2007-05-10  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* First beta released 0.2.0

2007-04-29  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Use getopt header with fosmount and fosread and add --help and
	  --version options.

2007-03-25  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* smascii: add an option to convert Carriage Return to Unix newline
	  (Line Feed).

	* Fix Ubuntu package builder with CDBS.

2007-03-24  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* Some fixes in all the project.

	* smascii: new tool to convert Smaky's text files to Extended ASCII
	  (ISO-8859-1) files.

2007-03-22  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: change all long variables for int because the FOS uses only
	  32 bits addresses. The lpow() function is removed and it was used only
	  for the c2l() function. Then a shift register is enough and faster.

2007-03-19  Mathieu Schroeter <mathieu.schroeter@gamesover.ch>

	* libfosfat: add a test for CHK value in BD and BL. If this value
	  changes, then NULL is returned and not the block.

tools/smascii.c
/*
 * FOS smascii: Smaky's ASCII to Extended ASCII (ISO-8859-1) converter
 * Copyright (C) 2007-2010 Mathieu Schroeter <mathieu@schroetersa.ch>
 *
 * Thanks to Pierre Arnaud for his help and the documentation
 *    And to Epsitec SA for the Smaky computers
 *
 * This file is part of Fosfat.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include "ascii.h"

#define BUFFER_SIZE 256

#define HELP_TEXT \
"Tool for convert Smaky text file to " \
"Extended ASCII (ISO-8859-1).\n\n" \
"Usage: smascii smaky_file converted_file [--unix]\n\n" \
" smaky_file            the smaky text file\n" \
" converted_file        the file converted\n" \
" --unix                the Carriage Return (Old Mac) will be\n" \
"                       converted to Line Feed (unix)\n" \
"\nPlease, report bugs to <mathieu@schroetersa.ch>.\n"

/*
 * Run the text conversion.
 *
 * input        Smaky text file
 * output       ISO-8859-1 text file
 * return 1 for success and 0 for error
 */
static int
run_conv (const char *input, const char *output, newline_t newline)
{
  int res = -1;
  FILE *in = NULL, *out = NULL;
  char buffer[BUFFER_SIZE];
  size_t lng;

  in = fopen (input, "rb");
  if (!in)
  {
    fprintf (stderr, "Reading error!\n");
    goto out;
  }

  out = fopen (output, "wb");
  if (!out)
  {
    fprintf (stderr, "Writing error!\n");
    goto out;
  }

  while ((lng = fread (buffer, 1, sizeof (buffer), in)))
  {
    size_t size;
    if (fos_sma2iso8859 (buffer, (unsigned int) lng, newline))
    {
      size = fwrite (buffer, 1, lng, out);
      if (size != lng)
      {
        fprintf (stderr, "Conversion error, bad size!\n");
        goto out;
      }
    }
    else
    {
      fprintf (stderr, "Conversion error!\n");
      goto out;
    }
  }

  printf ("File %s successfully converted to %s!\n", input, output);
  res = 0;

 out:
  if (in)
    fclose (in);
  if (out)
    fclose (out);

  return res;
}

static void
print_help (void)
{
  printf (HELP_TEXT);
}

int
main (int argc, char **argv)
{
  newline_t newline = ASCII_CR;

  if (argc < 3)
  {
    print_help ();
    return -1;
  }

  if (argc == 4 && !strcmp (argv[3], "--unix"))
    newline = ASCII_LF;

  return run_conv (argv[1], argv[2], newline);
}
